docs/language-reference/02-lexcial-structure.md

```diff
@@ -0,0 +1,114 @@
+Implementations *may* accept source units stored as files on disk, buffers in memory, or any appropriate implementation-specified means.
+
+Encoding
+--------
+
+When encoding is required, source units *should* be encoded using UTF-8.
+Implementations *may* support additional implemented-specified encodings.
+
+Whitespace
+----------
+
+_Horizontal whitespace_ consists of space (U+0020) and horizontal tab (U+0009).
+
+A _line break_ consists of a line feed (U+000A), carriage return (U+000D) or a carriage return followed by a line feed (U+000D, U+000A).
+Line breaks are used as line separators rather than terminators; it is not necessary for a source unit to end with a line break.
+
+Escaped Line Breaks
+-------------------
+
+An _escaped line break_ comprises a backslack (`\`, U+005C) follow immediately by a line break.
+
+Comments
+--------
+
+A _comment_ is either a line comment or a block comment.
+
+A _line comment_ comprises two forward slashes (`/`, U+002F) followed by zero or more characters that do not contain a line break.
+A line comment extends up to, but does not include, a subsequent line break or the end of the source unit.
+
+A _block comment_ begins with a forward slahs (`/`, U+002F) followed by an asterisk (`*`, U+0052). 
+A block comment is terminated by the next instance of an asterisk followed by a forward slash (`*/`).
+A block comment contains all characters between where it begins and where it terminates, including any line breaks.
+Block comments do not nest.
+It is an error if a block comment that begins in a source unit is not terminated in that source unit.
+
+Phases
+------
+
+Compilation of a source unit proceeds _as if_ the following steps are executed in order:
+
+1. Line numbering (for subsequent diagnostic messages) is noted based on the locations of line breaks
+
+2. Escaped line breaks are eliminated. No new characters are inserted to replace them. Any new escaped line breaks introduced by this step are not eliminated.
+
+3. All comments are replaced with a single space (U+0020)
+
+4. The source unit is _lexed_ into a sequence of tokens according the lexical grammar in this chapter
+
+5. The lexed sequence of tokens is _preprocessed_ to produce a new sequence of tokens (Chapter 3)
+
+6. Subsequent processing is performed on the preprocessed sequence of tokens
+
+Identifiers
+-----------
+
+An _identifier_ betgins with an upercase or lowercase ASCII leter (`A` through `Z`, `a` through `z`), or an underscore (`_`).
+After the first character, ASCII digits (`0` through `9`) may also be used in an identifier.
+
+The identifier consistent of a single underscore (`_`) is reserved by the language and must not be used by programs.
+Otherwise, there are no fixed keywords or reserved words.
+Words that name a built-in language construct can also be used as user-defined identifiers and will shadow the built-in definitions in the scope of their definition.
```

Comment:
Well, for right now this is a bit loose, but the short answer is that *yes* you can shadow a lot of the built-in language keywords.

From a mechanism standpoint: when the parser is trying to parse an expression, statement, or declaration and the lookahead token is an identifier, it attempts to look up that identifier in the current environment. If the lookup step finds a `SyntaxDecl` then it knows the identifier mapped to some kind of built-in syntax, and it will parse using the callback on that `SyntaxDecl`. If something in a more local scope shadows the same identifier, then lookup will find that declaration instead, and the parser won't parse using the `SyntaxDecl`.

There are a few pieces of rationale behind this design:

1. It is 100% desirable to have many keywords be usable as ordinary identifiers. This is both because HLSL has keywords with names like `triangle` and `sample` that are very likely to be things people want to use as local variables, *and* because Slang adds its own keywords (`extension`, `import`, etc.) that would make it impossible to be backwards-compatible with HLSL if users couldn't use those same identifiers to name functions/variables/etc.

2. It is super easy to work with and extend. Adding parsing logic a new language construct with a new keyword mostly amounts to adding a subroutine to do the parsing, plus registering its name/keyword at one place in the parser initialization. In comparison, approaches that recognize keywords during tokenization tend to require a lot more steps when adding a keyword.

3. Because syntax can be introduced as an ordinary kind of declaration, many of our keywords are just defined in the stdlib by naming the modifier keyword and the C++ AST node class that it should yield.

4. This flexibility can/will be required if we ever wanted to introduce a first-class hygienic macro system that allows user to introduce their own syntactic keywords.

I completely agree that this design choice could lead to confusion if people abuse it:

```hlsl
bool myFunc(bool true, bool false)
{
    // ... many lines later

    return true; // returns first parameter which might not be true
}

bool b = myFunc(false, true); // should return false
```

I can argue that people can always write confusing code when given flexible language rules (e.g., overloading has often been vilified for similar reasons), so this isn't *that* bad.

Still, the current realization of this feature has a lot of gotchas, like:

* The lookup rules that end up being used during parsing don't match those used for semantic checks. Perhaps most importantly, they don't work correctly for forward references (so if `true` is defined as a global variable later in your file, a reference to `true` earlier in the file won't necessarily see it).

* There are still some ad hoc rules in the parser that look for exact strings instead of always looking up `SyntaxDecl`s and these need to be cleaned up sooner or later (they are a maintenance headache).

I'd definitely be willing to consider changes that make some of our built-in syntax act more like keywords, perhaps by issuing an error on any attempt to use them as an ordinary identifier. We probably need to leave many of them as contextual keywords that can still be used to name functions/variables/etc.

I still maintain a dim hope of some day having a hygienic macro system in Slang, although I will freely admit that the longer we've been working on the language, the less demand I'm seeing for anything like that.